### PR TITLE
Add Path.mkdirs

### PR DESCRIPTION
**Pull request by patricoferris**
As per one of the tasks in #510. I'm almost certainly doing something wrong... Unfortunately I don't have a windows machine so I blindly added the windows test.

### CODE REVIEW

**Comment by patricoferris** [comment]
Hmm I think the Linux error (or the macOS/Windows incorrect success) is caused by catching https://github.com/ocaml-multicore/eio/blob/355f8dabf28b229dc18285971ebbce62e2c4f3f1/lib_eio_posix/fs.ml#L158 hence the diff in the mdx test :/

File: lib_eio/path.ml [hunk 1]
```diff
@@ -35,6 +35,27 @@ let mkdir ~perm ((t:#Fs.dir), path) =
     let bt = Printexc.get_raw_backtrace () in
     Exn.reraise_with_context ex bt "creating directory %a" pp (t, path)
 
+let mkdirs ?(exists_ok=false) ~perm ((t:#Fs.dir), path) =
+  let separate = String.split_on_char (Filename.dir_sep.[0]) in
```
Comment (talex5) [inline lib_eio/path.ml:39]: I think we decided in #124 that the path separator is always `/` in Eio paths.

File: lib_eio/path.ml [hunk 2]
```diff
@@ -35,6 +35,27 @@ let mkdir ~perm ((t:#Fs.dir), path) =
...
+      Exn.reraise_with_context ex bt "creating directory %a" pp (t, p)
+  in
+  let rec loop acc = function
+    | [] -> ()
+    | x :: xs ->
+      let dir = Filename.concat acc x in
+      make_directory dir;
+      loop dir xs
+  in
+  loop "" (separate path)
```
Comment (talex5) [inline lib_eio/path.ml:57]: It's probably more efficient to start by checking the direct parent. If that exists (which is the common) case, there's no need to check further up (see e.g. [Python's makedirs](https://github.com/python/cpython/blob/87de2fbb1e9b94f25ae8c0c233fb4690a13e4b02/Lib/os.py#L200) or [Bos's create](https://github.com/dbuenzli/bos/blob/6320f302eb9487518491511292a11ee6f04bd6e9/src/bos_os_dir.ml#L31)).

File: lib_eio/path.mli [hunk 3]
```diff
@@ -89,7 +89,11 @@ val with_open_out :
 (** {1 Directories} *)
 
 val mkdir : perm:File.Unix_perm.t -> _ t -> unit
-(** [mkdir ~perm t] creates a new directory [t] with permissions [perm]. *)
+(** [mkdir ?exist_ok ~perm t] creates a new directory [t] with permissions [perm]. *)
+
+val mkdirs : ?exists_ok:bool -> perm:File.Unix_perm.t -> _ t -> unit
+(** Recursively create directories. If [exist_ok] is [false] (the default) then we raise
```
Comment (talex5) [inline lib_eio/path.mli:95]: `exists_ok` should only affect the final directory, not the intermediate ones (most of which will exist).

File: lib_eio/path.ml (hunk 1 above)
Comment (patricoferris) [reply lib_eio/path.ml]: Oh right, thanks for pointing that out :))

**Comment by talex5** [comment]
Closing in favour of the newer version in #625 - thanks!

**Pull request closed without merging**